[WFS] Improve support of layers with unknown geometry type with Deegree servers

rouault opened the pull request:
or any server implementing IsPoint, IsCurve and IsSurface filtering functions.
    
When such a layer is added to the project, the Select Items to Add
dialog box used for example for OGR vector layers with unknown geometry
type is displayed, enabling the user to select one or several layers
among ones of NoGeometry/MultiPoint/MultiCurve/MultiSurface/GeometryCollection
types.

![image](https://user-images.githubusercontent.com/1192433/206298574-33a67922-e0c5-4a8b-ae2e-b90923399c22.png)

Fixes #49328

Funded by QGIS-DE user group (QGIS Anwendergruppe Deutschland e.V.)

rouault committed:
Add QgsGeometry::convertToCurvedMultiType()

rouault committed:
[WFS] Improve support of layers with unknown geometry type with Deegree servers

or any server implementing IsPoint, IsCurve and IsSurface filtering
functions.

When such a layer is added to the project, the Select Items to Add
dialog box used for example for OGR vector layers with unknown geometry
type is displayed, enabling the user to select one or several layers
among ones of NoGeometry/MultiPoint/MultiCurve/MultiSurface/GeometryCollection
types.

Fixes #49328

rouault commented:
> Is this intended?

It is the default behaviour of the application component of QGIS. It should be patched somehow if one wanted to change it. That said, it looks to me that it may be useful as a warning for users that they're going to add a layer that is perhaps not going to behave properly in QGIS due to potential mix of geometry types in it.

pathmapper commented on src/providers/wfs/qgswfsprovidermetadata.cpp:
```diff
@@ -0,0 +1,365 @@
+        connect( thisRequest, &QgsWfsRequest::downloadFinished, thisRequest, downloadFinishedLambda );
+      }
+      QgsSettings s;
+      const double timeout = s.value( QStringLiteral( "qgis/wfsGetFeatureGeometryTypesTimeout" ), 2.0 ).toDouble();
```
I can't spot `qgis/wfsGetFeatureGeometryTypesTimeout` in advanced settings to customize it. It's also not listed in QGIS3.ini.

rouault replied:
addressed in 761f0292597d70118145a7aed0857d456a811099

pathmapper replied:
:+1:

pathmapper commented:
> it may be useful as a warning for users that they're going to add a layer that is perhaps not going to behave properly in QGIS due to potential mix of geometry types in it

Agree with your thoughts. I would expect that the dialog will show often, because I've got the impression that in DescribeFeatureType the geometry is often advertised as `gml:GeometryPropertyType` instead of a distinct geometry type but maybe making people more aware of this by showing the dialog will change this habit a bit :smile: .

In the dialog the checkbox "Add layers to a group" currently has no effect, maybe the name of the FeatureType could be used for the group name.

![image](https://user-images.githubusercontent.com/20856381/206842317-03a095d5-08f6-4d68-9e3b-323737821d5b.png)

rouault committed:
[WFS provider] Write qgis/wfsGetFeatureGeometryTypesTimeout settings key if it does not exist

rouault committed:
Add a QgsProviderMetadata:suggestGroupNameForUri() method and implement it for WFS to have WFS sublayers put in a group

rouault commented:
> In the dialog the checkbox "Add layers to a group" currently has no effect, maybe the name of the FeatureType could be used for the group name.

Adressed in 78c9aa70c7d2c17acbfd0c964b3e27f8b5234967

![image](https://user-images.githubusercontent.com/1192433/206863086-27d790a9-1e1e-42ef-a1d3-4c04aecda2c6.png)

pathmapper commented:
> Adressed in https://github.com/qgis/QGIS/commit/78c9aa70c7d2c17acbfd0c964b3e27f8b5234967

Thanks! 

What about keeping the name of the FeatureType **also** for the layer names (like it was before)? Adding multiple layers named e.g. only `MultiPoint` from different FeatureTypes and changing layer order while moving layers outside their _original_ group could lead to confusion or a need to manually rename the layers to include the FeatureType name to keep track of things.

rouault committed:
[WFS provider] Include typename in sublayers name

rouault commented:
> What about keeping the name of the FeatureType **also** for the layer names (like it was before)?

Done in d533e67346

![image](https://user-images.githubusercontent.com/1192433/206873421-9e240ca9-0b22-435c-a138-1f3c8e4b9ef7.png)

pathmapper commented:
QGIS crashes if `qgis/wfsGetFeatureGeometryTypesTimeout` = 0 when adding a FeatureType, e.g. `xplan:BP_AnpflanzungBindungErhaltung`:

![image](https://user-images.githubusercontent.com/20856381/206891734-aaf1b323-75ec-482b-aff2-f00de49276c9.png)

pathmapper commented:
With this PR, if any FeatureType from https://www.wfs.nrw.de/geobasis/wfs_nw_dvg is added to the project, an error like the following is logged to the console:

```shell
$ ./output/bin/qgis
ERROR: Status 2: File  pagingEnabled='false' preferCoordinatesForWfsT11='false' restrictToRequestBBOX='1' srsname='EPSG:25832' typename='dvg:nw_dvg1_bld' url='https://www.wfs.nrw.de/geobasis/wfs_nw_dvg' version='auto' could not be found
```

rouault committed:
QgsWfsProviderMetadata::querySublayers(): fix crash if wfsGetFeatureGeometryTypesTimeout setting set to 0

rouault committed:
QgsProviderSublayerTask: make it possible to restrict it to a particular provider

Avoids the 'ERROR: Status 2: File  pagingEnabled='false'
preferCoordinatesForWfsT11='false' restrictToRequestBBOX='1'
srsname='EPSG:25832' typename='dvg:nw_dvg1_bld'
url='https://www.wfs.nrw.de/geobasis/wfs_nw_dvg' version='auto' could
not be found' type of message in the scenario described by
https://github.com/qgis/QGIS/pull/51144#issuecomment-1345507627

rouault commented:
> QGIS crashes if `qgis/wfsGetFeatureGeometryTypesTimeout` = 0 

fixed per a2195a9c6385f9301fd3aae80d5517191d9da9a4

> With this PR, if any FeatureType from https://www.wfs.nrw.de/geobasis/wfs_nw_dvg is added to the project, an error like the following is logged to the console:

I had a hard time figuring out that. The error message comes from a unidentifier provider. I've fixed / workaround this by 33f0ba134980a9addc5cf23b5841452449022a2a

rouault committed:
[WFS]: QgsWfsProviderMetadata::querySublayers(): do not issue GetCapabilities request on non-WFS URIs

pathmapper commented on src/providers/wfs/qgswfsprovidermetadata.cpp:
```diff
@@ -0,0 +1,384 @@
+  doc.appendChild( filterElem );
+  QString filterPrefix( caps.version.startsWith( "2.0" ) ? QStringLiteral( "fes" ) : QStringLiteral( "ogc" ) );
+
+  QDomElement andElem = doc.createElement( filterPrefix + QStringLiteral( ":And" ) );
```
This `And` element includes only the `!PropertyIsNull` filter but not the `!IsPoint`, `!IsCurve`, `!IsSurface` filters, resulting in an [invalid filter](https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=2.0.0&TYPENAMES=xplan:BP_AnpflanzungBindungErhaltung&COUNT=1&SRSNAME=EPSG:25832&FILTER=%3Cfes:Filter%20xmlns:fes%3D%22http://www.opengis.net/fes/2.0%22%3E%0A%20%3Cfes:And%3E%0A%20%3Cfes:Not%3E%0A%20%3Cfes:PropertyIsNull%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:PropertyIsNull%3E%0A%20%3C/fes:Not%3E%0A%20%3C/fes:And%3E%0A%20%3Cfes:PropertyIsEqualTo%3E%0A%20%3Cfes:Function%20name%3D%22IsPoint%22%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:Function%3E%0A%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%3C/fes:PropertyIsEqualTo%3E%0A%20%3Cfes:PropertyIsEqualTo%3E%0A%20%3Cfes:Function%20name%3D%22IsCurve%22%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:Function%3E%0A%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%3C/fes:PropertyIsEqualTo%3E%0A%20%3Cfes:PropertyIsEqualTo%3E%0A%20%3Cfes:Function%20name%3D%22IsSurface%22%3E%0A%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%3C/fes:Function%3E%0A%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%3C/fes:PropertyIsEqualTo%3E%0A%3C/fes:Filter%3E%0A).

rouault replied:
good catch. sorry for that. fixed per e1c660dd1c
Now manual testing of ``vl = QgsVectorLayer("pagingEnabled='true' preferCoordinatesForWfsT11='false' restrictToRequestBBOX='1' srsname='EPSG:25832' typename='xplan:BP_AnpflanzungBindungErhaltung' url='https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs' version='auto' geometryTypeFilter='GeometryCollection'", 'test', 'WFS')`` results in the correct https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=2.0.0&TYPENAMES=xplan:BP_AnpflanzungBindungErhaltung&COUNT=1&SRSNAME=EPSG:25832&FILTER=%3Cfes:Filter%20xmlns:fes%3D%22http://www.opengis.net/fes/2.0%22%3E%0A%20%3Cfes:And%3E%0A%20%20%3Cfes:Not%3E%0A%20%20%20%3Cfes:PropertyIsNull%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:PropertyIsNull%3E%0A%20%20%3C/fes:Not%3E%0A%20%20%3Cfes:PropertyIsEqualTo%3E%0A%20%20%20%3Cfes:Function%20name%3D%22IsPoint%22%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:Function%3E%0A%20%20%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%20%3C/fes:PropertyIsEqualTo%3E%0A%20%20%3Cfes:PropertyIsEqualTo%3E%0A%20%20%20%3Cfes:Function%20name%3D%22IsCurve%22%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:Function%3E%0A%20%20%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%20%3C/fes:PropertyIsEqualTo%3E%0A%20%20%3Cfes:PropertyIsEqualTo%3E%0A%20%20%20%3Cfes:Function%20name%3D%22IsSurface%22%3E%0A%20%20%20%20%3Cfes:ValueReference%3Eposition%3C/fes:ValueReference%3E%0A%20%20%20%3C/fes:Function%3E%0A%20%20%20%3Cfes:Literal%3Efalse%3C/fes:Literal%3E%0A%20%20%3C/fes:PropertyIsEqualTo%3E%0A%20%3C/fes:And%3E%0A%3C/fes:Filter%3E%0A request being emitted

pathmapper replied:
Thanks and no worries Even, I'm happy to give some useful feedback and still impressed what you've created here.

rouault committed:
[WFS] Fix filter for geometryTypeFilter='GeometryCollection' layers

pathmapper commented:
When layers are added without group, the layername could be shorter (currently 2x FeatureType included):
![image](https://user-images.githubusercontent.com/20856381/206924111-e2af7545-4ffe-41aa-93c1-02615037e867.png)

pathmapper commented:
There's still a `count=1` request when adding a layer via the dialog - we don't need this request and could instead request directly with the appropriate filter to receive all features, right?

![image](https://user-images.githubusercontent.com/20856381/206924397-0234c888-6596-4548-978a-24ced8cf0410.png)


![image](https://user-images.githubusercontent.com/20856381/206924321-4c4517ba-9166-4ec6-a9e8-647dfa87d263.png)

rouault committed:
QgsAppLayerHandling::addSublayers(): avoid prefixing layer name with basename if it starts with basename

Adresses https://github.com/qgis/QGIS/pull/51144#issuecomment-1345634678

rouault commented:
> When layers are added without group, the layername could be shorter (currently 2x FeatureType included):

addressed per 8d02bf27a0

> There's still a count=1 request when adding a layer via the dialog - we don't need this request and could instead request directly with the appropriate filter to receive all features, right?

Unfortunately, it is needed to know if there are gml:description, gml:name, gml:identifier fields. Cf  #42660 / ec96b494901006b01df7a3c324415aa7189a2e40

pathmapper commented:
> it is needed to know if there are gml:description, gml:name, gml:identifier fields

It seems that there are services (confirmed with 2 different instances of xPlanBox) where these fields are included (named a bit differently) in the DescripeFeatureType response.

Maybe in such case we could get the information from DescribeFeatureType and skip the count=1?

https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs?SERVICE=WFS&REQUEST=DescribeFeatureType&VERSION=2.0.0&TYPENAMES=xplan:BP_AnpflanzungBindungErhaltung&TYPENAME=xplan:BP_AnpflanzungBindungErhaltung

![grafik](https://user-images.githubusercontent.com/20856381/207538668-0ecd018e-5c85-4256-8fa0-16e53d73f8d0.png)

rouault committed:
[WFS provider] Do not emit initial COUNT=1 GetFeature request if DescribeFeatureType advertizes gmlId, gmlName and gmlDescription fields

rouault commented:
> Maybe in such case we could get the information from DescribeFeatureType and skip the count=1?

done in https://github.com/qgis/QGIS/pull/51144/commits/9dfb017b2bec4f5afbce456d28935a09fa2166ea . It seems that this was a workaround for clients like QGIS before discovery of gml:name/identifier/description was implemented :-)

pathmapper commented:
Yep, this might be the reason :) Thanks for implementing this!

rouault commented:
@pathmapper any testing left on your side before we ask a committer to review ?

nyalldawson commented on src/core/geometry/qgsgeometry.h:
```diff
@@ -2376,6 +2376,21 @@ class CORE_EXPORT QgsGeometry
+     * the current geometry is a multipoint or a geometry collection. A single
+     * point will be transformed to a multipoint.
+     *
+     * \returns TRUE in case of success and FALSE else
```
\since QGIS 3.30

rouault replied:
done

nyalldawson commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1590,6 +1596,52 @@ bool QgsGeometry::convertToMultiType()
+  {
+    for ( int i = 0; i < sourceMultiGeom->numGeometries(); ++i )
+    {
+      if ( !multiGeom->addGeometry( sourceMultiGeom->geometryN( i )->clone() ) )
```
Potentially we could avoid a bunch of cloning here if we moved the guts of this to the geometry subclasses (e.g. a multilinestring could just transfer the line parts into the compound curve result parts), but we can defer that until this method sees more widespread use.

rouault replied:
agreed

nyalldawson commented on src/core/providers/qgsprovidersublayertask.h:
```diff
@@ -55,6 +55,12 @@ class CORE_EXPORT QgsProviderSublayerTask : public QgsTask
 
+    /**
+     * Constructor for QgsProviderSublayerTask, which retrieves sublayer details for the
+     * specified \a uri, restricted to a particular provider
+     */
```
```suggestion
     * specified \a uri, restricted to a particular provider.
     *     
     * \since QGIS 3.30
     */
```

rouault replied:
done

nyalldawson commented on src/app/qgsprovidersublayersdialog.cpp:
```diff
@@ -112,21 +112,31 @@ void QgsProviderSublayerDialogModel::setGeometryTypesResolved( bool resolved )
+
+  if ( !fileName.isEmpty() )
+  {
+    mGroupName = fileName;
+    const QgsSettings settings;
+    if ( settings.value( QStringLiteral( "qgis/formatLayerName" ), false ).toBool() )
+    {
+      mGroupName = QgsMapLayer::formatLayerName( mGroupName );
+    }
```
setGroupName( fileName );

rouault replied:
done

nyalldawson commented on tests/src/python/test_qgsgeometry.py:
```diff
@@ -2867,6 +2867,85 @@ def testConvertToMultiType(self):
+
+        multiline = QgsGeometry.fromWkt('MultiLineString ((1 0, 2 0))')
+        assert multiline.convertToCurvedMultiType()
+        expWkt = 'MultiCurve (LineString (1 0, 2 0))'
```
should this be MultiCurve( CompoundCurve( ) ) ? That's what I would have immediately expected from this function.

rouault replied:
a LineString is a legal component in a MultiCurve, since it is a Curve.

nyalldawson replied:
I realise that -- it's just that the method name "convert to curved ... type" immediately suggests to me that the result would be an explicit curved geometry type. (Like if we had a non-promote-to-multi version of this I'd expect feeding a single linestring would result in a compound curve output.)

rouault replied:
the use case here is that the top "geometry" must be a curved multi geometry type. e.g to be able to export a WFS layer mixing linestring, compoundcurve, multilinestring, etc. as a multicurve one for formats like GeoPackage or PostGIS that strictly enforce the "top" geometry type. The inner type doesn't matter much, as long as it is legal for the container.

rouault replied:
not necessarily, the ultimate source of truth, but the behaviour I implemented is similar to OGR ForceTo(wkbMultiCurve):
```
>>> g = ogr.CreateGeometryFromWkt('LINESTRING(0 0,1 1)')
>>> g = ogr.ForceTo(g, ogr.wkbMultiCurve)
>>> g.ExportToIsoWkt()
'MULTICURVE ((0 0,1 1))'
>>> g.GetGeometryRef(0).ExportToIsoWkt()
'LINESTRING (0 0,1 1)'
```

nyalldawson commented on tests/src/python/test_qgsgeometry.py:
```diff
@@ -2867,6 +2867,85 @@ def testConvertToMultiType(self):
+
+        poly = QgsGeometry.fromWkt('Polygon ((1 0, 2 0, 2 1, 1 1, 1 0))')
+        assert poly.convertToCurvedMultiType()
+        expWkt = 'MultiSurface (Polygon((1 0, 2 0, 2 1, 1 1, 1 0)))'
```
(similarly, I'd have expected MultiSurface(CurvePolygon()) )

rouault replied:
same as above

nyalldawson commented on src/providers/wfs/qgswfsprovidermetadata.cpp:
```diff
@@ -0,0 +1,384 @@
+
+    const auto downloaderLambda = [&]()
+    {
+      QEventLoop loop;
```
I'd capture feedback here and support early cancellation via it.

rouault replied:
done

nyalldawson commented on src/core/providers/qgsprovidermetadata.h:
```diff
@@ -457,6 +457,16 @@ class CORE_EXPORT QgsProviderMetadata : public QObject
     virtual QList< QgsProviderSublayerDetails > querySublayers( const QString &uri, Qgis::SublayerQueryFlags flags = Qgis::SublayerQueryFlags(), QgsFeedback *feedback = nullptr ) const;
 
+    /**
+     * Returns a name that can be used as a group name for sublayers got from
```
```suggestion
     * Returns a name that can be used as a group name for sublayers retrieved from
```

rouault replied:
done

pathmapper commented:
I'm done with testing from my side, functionality LGTM :+1: 

There's one last thing which maybe could be optimized:

After selecting which layer(s) to load from the dialog, according to the network log there are additional network requests with `GetCapabilities` and `DescribeFeatureType` which maybe aren't needed at this stage:

https://xplanung.freiburg.de/xplansyn-wfs/services/xplansynwfs

![image](https://user-images.githubusercontent.com/20856381/207802203-b99a77e2-0dde-4e9b-ae48-d86b6f65c7b2.png)

---

Also for a service which doesn't offer the IsPoint, Is Curve, IsSurface functions, with this PR there are more requests (5 from network) compared to QGIS 3.28.1 (3 from network):

Additional network requests are `GetFeature` with count=1 and `DescribeFeatureType`

https://www.wfs.nrw.de/geobasis/wfs_nw_dvg

**This PR:**
![image](https://user-images.githubusercontent.com/20856381/207803246-e07c7514-5113-4243-b761-243d0ba43900.png)



**QGIS 3.28.1**
![image](https://user-images.githubusercontent.com/20856381/207804323-c52832af-3ffe-48c8-8955-d0df078420bd.png)

rouault committed:
[WFS provider] Fix getCachedCapabilities() short-lived in-memory cache for WFS capabilities

rouault committed:
[WFS provider] Add short-lived in-memory cache for DescribeFeatureType and GetFeature COUNT=1 queries

rouault commented:
> There's one last thing which maybe could be optimized:

Addressed per 60e487d673e7e327a3703e368f348ebb68983bee and 1d4c34faa4b8199051a9927122a0379b0b449769.

pathmapper commented:
Thanks, with this change, now if one select all three layers in the dialog without adding them to a group, only one layer is added:

https://www.wfs.nrw.de/geobasis/wfs_nw_dvg
![image](https://user-images.githubusercontent.com/20856381/207876759-0be04153-b107-4945-9667-959edee33e2c.png)

![image](https://user-images.githubusercontent.com/20856381/207877754-5e645f8b-b495-4a76-a177-1d62470b7ef2.png)

---

And for https://www.wfs.nrw.de/geobasis/wfs_nw_dvg the count=1 is still emitted twice:

![image](https://user-images.githubusercontent.com/20856381/207883808-0779a705-bd29-4e54-ae41-e63671e509c2.png)

rouault committed:
[WFS provider] Avoid console warning when feedback == nullptr

rouault commented:
> with this change, now if one select all three layers in the dialog without adding them to a group, only one layer is added:

I don't reproduce this. Layer addition works consistently for me, with or without a group. Are you sure your build is OK ? I've force-pushed commits yesterday, so you need to do something like "git fetch rouault; git reset --hard rouault/wfs_sublayers" assuming rouault is the name for my remote

> And for https://www.wfs.nrw.de/geobasis/wfs_nw_dvg the count=1 is still emitted twice

hum I see actually this query is emitted in asynchronous mode, and it is tricky to make it work with the cache (I've just tried, and it results in hangs). I won't tackle this extra optimization.

pathmapper commented:
> I don't reproduce this

Can't produce neither with a totally fresh build, sorry.

> won't tackle this extra optimization

Sure, thanks for giving it a quick try.

---

As said, all testing done from my side. Thanks a lot for addressing all the feedback. I'm very happy about the result! :tada: :tada: :tada: 

Also thanks to @nyalldawson for the review.

rouault commented:
keep alive

rouault commented:
@nyalldawson I believe the last remaining point of discussion was related to convertToCurvedMultiType() behavior. Is the behaviour I implemented currently acceptable ?

rouault commented:
@nyalldawson gentle ping w.r.t above question :-)

nyalldawson commented on python/gui/auto_generated/qgsprovidersublayersdialog.sip.in:
```diff
@@ -56,6 +56,7 @@ Dialog for selecting provider sublayers.
   public:
 
     QgsProviderSublayersDialog( const QString &uri,
+                                const QString &providerKey,
```
Sorry - I should have picked this one up earlier, but adding this argument here is an API break. Can you add it as an optional argument at the end to avoid this?

nyalldawson replied:
Whoops I see this class was introduced in 3.30... so ignore my comment, still on holiday brain!

nyalldawson commented:
@rouault sorry, been on holiday break! I've just approved but with the above comment re api break
